Add contract extensibility docs (#140)
OpenZeppelin/cairo-contracts: docs/Extensibility.md

# Extensibility

> Expect this pattern to evolve (as it has already done) or even disappear if [proper extensibility features](https://community.starknet.io/t/contract-extensibility-pattern/210/11?u=martriay) are implemented into Cairo.

* [The extensibility problem](#the-extensibility-problem)
* [The pattern ™️](#the-pattern)
  * [Libraries](#libraries)
  * [Contracts](#contracts)
* [Presets](#presets)
* [Emulating hooks](#emulating-hooks)

## The extensibility problem

> Smart contract development is a critical task. As all software development, is error prone; but unlike most scenarios a bug can result in major losses for organizations as well as individuals. Therefore writing complex smart contracts is a delicate task.

Reviewer: I think this reads a little better
```suggestion
Smart contract development is a critical task. As with all software development, it is error prone; but unlike most scenarios, a bug can result in major losses for organizations as well as individuals. Therefore writing complex smart contracts is a delicate task.
```

One of the best approaches to minimize introducing bugs is to reuse existing, battle-tested code, a.k.a. using libraries. But code reutilization in StarkNet’s smart contracts is not easy:

- Cairo has no explicit smart contract extension mechanisms such as inheritance or composability

> - There’s no function overloading nor aliasing, making function selector collisions very likely – more so considering selectors do not take function arguments into account

Reviewer: ```suggestion
- There’s no function overloading nor aliasing which makes function selector collisions very likely – more so considering selectors do not take function arguments into account
```

> - There’s no function overloading nor aliasing which makes function selector collisions very likely – more so considering selectors do not take function arguments into account

Author: Yes thank you, I meant [overriding as I wrote below](https://github.com/OpenZeppelin/cairo-contracts/pull/140/files#diff-27ec49df67314bc4ba62e12685dcb20908f3cdbf292a8ce08e066d76afba8259R65)

- Any `@external` function defined in an imported module will be automatically re-exposed by the importer (i.e. the smart contract)

> To overcome these problems, this library builds on the following guidelines™.

Reviewer: Correct me if I'm wrong, but I believe this references the library of libraries here. I wonder if we should use a different name (e.g. collection of libraries, library of contracts). I think something akin to the suggestion adds clarity.
```suggestion
To overcome these problems, our StarkNet library builds on the following guidelines™.
```

Author: Good catch. "this repository"?

Reviewer: "this repository" might be appropriate. There's also "this protocol"

> To overcome these problems, this library builds on the following guidelines™.

Author: ```suggestion
To overcome these problems, this project builds on the following guidelines™.
```

## The pattern

> The idea is to have two types of cairo files: libraries and contracts. The flow then is for libraries to define reusable logic and storage variables which can then be extended and exposed by contracts. Contracts can be deployed, libraries cannot.

Reviewer: ```suggestion
The idea is to have two types of Cairo modules: libraries and contracts. The flow then is for libraries to define reusable logic and storage variables which can then be extended and exposed by contracts. Contracts can be deployed, libraries cannot.
```

> To minimize risk, boilerplate, and avoid semantic clashes, we follow these rules:

Author: Function names clashing.

```suggestion
To minimize risk, boilerplate, and avoid function naming clashes, we follow these rules:
```

### Libraries

- All function and storage variable names must be prefixed with the file name to prevent clashing with other libraries (e.g. `ERC20_approve` in the `ERC20` library)
- Must not implement any `@external` or `@view` functions
- Must not implement constructors
- Must not call initializers on any function

> - Should implement initializers if needed (as any other library function, never as `@external`)

Author: ```suggestion
- Should implement initializer functions to mimic construction logic if needed (as any other library function, never as `@external`)
```

### Contracts

- Can import from libraries

> - Should implement external functions if needed

Reviewer: For consistency
```suggestion
- Should implement `@external` functions if needed
```

- Should implement a constructor that calls initializers

> - Must not call initializers on any function aside the constructor

Reviewer 2: ```suggestion
- Must not call initializers in any function beside the constructor
```

> Note that since initializers will never be marked as `@external` and they won’t be called from anywhere but the contract constructor, there’s no risk of double-initialization. It’s up to the library developers not to make initializers inter-dependent to avoid weird dependency paths.

Reviewer 2: This is contradictory. There is risk of double initialization.

```suggestion
Note that since initializers will never be marked as `@external` and they won’t be called from anywhere but the contract constructor, there’s no risk of re-initialization after deployment. It’s up to the library developers not to make initializers interdependent to avoid weird dependency paths that may lead to double initialization of libraries.
```

  - [ERC20_Mintable](https://github.com/OpenZeppelin/cairo-contracts/blob/main/contracts/token/ERC20_Mintable.cairo)
  - [ERC20_Pausable](https://github.com/OpenZeppelin/cairo-contracts/blob/main/contracts/token/ERC20_Pausable.cairo)
- [ERC165](https://github.com/OpenZeppelin/cairo-contracts/blob/main/contracts/ERC165.cairo)
- [ERC721](https://github.com/OpenZeppelin/cairo-contracts/blob/main/contracts/token/ERC721.cairo)
  - [ERC721_Enumerable](https://github.com/OpenZeppelin/cairo-contracts/blob/main/contracts/token/ERC721_Enumerable.cairo)
  - [ERC721_Mintable](https://github.com/OpenZeppelin/cairo-contracts/blob/main/contracts/token/ERC721_Mintable.cairo)
  - [ERC721_Pausable](https://github.com/OpenZeppelin/cairo-contracts/blob/main/contracts/token/ERC721_Pausable.cairo)

## Emulating hooks

Unlike the Solidity version of [OpenZeppelin Contracts](https://github.com/OpenZeppelin/openzeppelin-contracts), this library does not implement [hooks](https://docs.openzeppelin.com/contracts/4.x/extending-contracts#using-hooks). The main reason being that Cairo does not support overriding functions.

> This is how a hook looks like in Solidity:

Reviewer: ```suggestion
This is what a hook looks like in Solidity:
```

```js
abstract contract ERC20Pausable is ERC20, Pausable {

>     /**
>      * @dev See {ERC20-_beforeTokenTransfer}.
>      *
>      * Requirements:
>      *
>      * - the contract must not be paused.
>      */

Reviewer 2: I feel like this doesn't add anything here.

```suggestion
```

Author: Agree it was a full snippet copypaste

>     function _beforeTokenTransfer(
>         address from,
>         address to,
>         uint256 amount
>     ) internal virtual override {

Reviewer 2: Makes it easier to read IMO:

```suggestion
    function _beforeTokenTransfer(address from, address to, uint256 amount) internal virtual override {
```

Author: Your own code ¯\\_(ツ)_/¯

Reviewer 2: Different context.

        super._beforeTokenTransfer(from, to, amount);

        require(!paused(), "ERC20Pausable: token transfer while paused");
    }
}
```

> Instead, using the extensibility pattern we can simply extend the library implementation of a function (namely `transfer`) by adding lines before or after calling it. This way, we can get away with:

Reviewer: ```suggestion
Instead, the extensibility pattern allows us to simply extend the library implementation of a function (namely `transfer`) by adding lines before or after calling it. This way, we can get away with:
```

```python
@external
func transfer{
        syscall_ptr : felt*, 
        pedersen_ptr : HashBuiltin*,
        range_check_ptr
    }(recipient: felt, amount: Uint256) -> (success: felt):
    Pausable_when_not_paused()

>     ERC20_transfer(recipient, amount)

Reviewer: Great job with the comparison!
